File: 0125-valid-palindrome/0125-valid-palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        '''
        U:
            q) input str can have any letter?
            
        M: two ptrs
        
        P:
            1. take care of edge cages (empty, one letter)
            2. init two val for left and right
            3. while left < right, check whether each letter is lowercase letter and non-alphabetical letter
            4. in the loop, check whether each letter is the same. if it is, then left++, right--
        '''
        
        left = 0
        right = len(s)-1
        
        while left < right:
            while left < right and not s[left].isalnum():
                left +=1
            while left < right and not s[right].isalnum():
                right-=1
                
            if left < right and s[left].lower() != s[right].lower():
                return False
            
            left+=1
            right-=1
            
        return True

File: 0125-valid-palindrome/0125-valid-palindrome.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    /*
    - remove non-alphabenumric. only contain number and lowercase.
    - empty str => true
    
    1. check edge cases(null, one length)
    2. init two ptrs(left and right)
    3. convert the input str to contain numbers and lowercases only
    4. while left <= right
        3.1) if left element and right element are same, left++, right--
        3.2) else return false
    5. return true
    */
 
    if(s.length === 0 ) return true;
    
    const lowercases = s.toLowerCase();
    const removeSpaces = lowercases.split(" ").join("");
    const onlyAlphabets = removeSpaces.replace(/[^a-zA-Z0-9 ]/g, '');
    
    let left = 0;
    let right = onlyAlphabets.length-1;
    while(left <= right){
        console.log(onlyAlphabets[left])
        console.log(onlyAlphabets[right])
        if(onlyAlphabets[left] !== onlyAlphabets[right]) return false;
        left++;
        right--
    } return true
}